pkg/telemetry/vector.go

```diff
@@ -0,0 +1,181 @@
+			vectorImage := defaultVectorImage
+			if telemetrySpec.Vector != nil && telemetrySpec.Vector.Image != "" {
+				vectorImage = telemetrySpec.Vector.Image
+			}
+
+			// Create the definition of the Vector agent container
+			vectorAgentContainer := corev1.Container{
+				Name:            cassandra.VectorContainerName,
+				Image:           vectorImage,
+				ImagePullPolicy: corev1.PullIfNotPresent,
+				Env: []corev1.EnvVar{
+					{Name: "VECTOR_CONFIG", Value: "/etc/vector/vector.toml"},
+					{Name: "VECTOR_ENVIRONMENT", Value: "kubernetes"},
+					{Name: "VECTOR_HOSTNAME", ValueFrom: &corev1.EnvVarSource{FieldRef: &corev1.ObjectFieldSelector{FieldPath: "spec.nodeName"}}},
+				},
+				VolumeMounts: []corev1.VolumeMount{
+					{Name: "vector-config", MountPath: "/etc/vector"},
+				},
+				Resources: vectorContainerResources(telemetrySpec),
+			}
+
+			logger.Info("Updating Vector agent in Cassandra pods")
+			cassandra.UpdateVectorContainer(&dcConfig.PodTemplateSpec, func(container *corev1.Container) {
+				*container = vectorAgentContainer
+			})
+
+			// Create the definition of the Vector agent config map volume
+			vectorAgentVolume := corev1.Volume{
+				Name: "vector-config",
+				VolumeSource: corev1.VolumeSource{
+					ConfigMap: &corev1.ConfigMapVolumeSource{
+						LocalObjectReference: corev1.LocalObjectReference{Name: VectorAgentConfigMapName(k8cName)},
+					},
+				},
+			}
+
+			cassandra.AddVolumesToPodTemplateSpec(dcConfig, vectorAgentVolume)
+		}
+	}
+
+	return nil
+}
+
+func vectorContainerResources(telemetrySpec *telemetry.TelemetrySpec) corev1.ResourceRequirements {
+	if telemetrySpec.Vector.Resources != nil {
+		return *telemetrySpec.Vector.Resources
+	}
+
+	return corev1.ResourceRequirements{
+		Requests: corev1.ResourceList{
+			corev1.ResourceCPU:    resource.MustParse(DefaultVectorCpuRequest),
+			corev1.ResourceMemory: resource.MustParse(DefaultVectorMemoryRequest),
+		},
+		Limits: corev1.ResourceList{
+			corev1.ResourceMemory: resource.MustParse(DefaultVectorMemoryLimit),
+			corev1.ResourceCPU:    resource.MustParse(DefaultVectorCpuLimit),
+		},
+	}
+}
+
+type VectorConfig struct {
+	Sinks          string
+	ScrapePort     int32
+	ScrapeInterval int32
+}
+
+func CreateCassandraVectorToml(ctx context.Context, telemetrySpec *telemetry.TelemetrySpec, dcConfig *cassandra.DatacenterConfig, remoteClient client.Client, namespace string) (string, error) {
+	sinks := `
+[sinks.console]
+type = "console"
+inputs = [ "cassandra_metrics" ]
+target = "stdout"
+
+  [sinks.console.encoding]
+  codec = "json"`
+
+	if telemetrySpec.Vector.Config != nil {
+		// Read the Vector provided config map content and use it as the Vector sink config
+		vectorConfigMap := &corev1.ConfigMap{}
+		err := remoteClient.Get(ctx, types.NamespacedName{Namespace: namespace, Name: telemetrySpec.Vector.Config.Name}, vectorConfigMap)
+		if err != nil {
+			return "", err
+		} else {
+			sinks = vectorConfigMap.Data["vector.toml"]
```

Comment:
You might want to test whether the key exists, otherwise you would be assigning an empty string to `sinks`.

Reply:
good point 👍